include/fiddle/interaction/interaction_utilities.h

```diff
@@ -260,6 +263,155 @@ namespace fdl
                        const Vector<double>         &position,
                        const Vector<double>         &spread_values);
 
+  /**
+   * Compute intersection the point of a line with an edge
+   *
+   * WARNING: This code is specialized to the case in which q is a unit vector
+   * aligned with the coordinate axes.
+   *
+   * Linear interpolation:
+   *
+   * 0.5*(1-u)*p0 + 0.5*(1+u)*p1 = r + t*q
+   *
+   * Factor the interpolation formula:
+   *
+   * 0.5*(p1-p0)*u+0.5*(p1+p0) = r + t*q
+   *
+   * Solve for u:
+   *
+   *    a*u + b = 0
+   *
+   * with:
+   *
+   *   a = 0.5*(-p0+p1)
+   *   b = 0.5*(p0+p1) - r
+   * @param[out] t_vals The vector containing the intersection point in Eulerian
+   * coordinate and element reference coordinate.
+   *
+   * @param[in] element_vertices The array containing Eulerian location of
+   * vertices of the given element.
+   *
+   * @param[in] r The center location of the FD stencil`.
+   *
+   * @param[in] q a unit vector aligned with the coordinate axes.
+   *
+   * @param[in] tol a given tolerance for deciding if the intersection point is
+   * within the edge.
+   *
+   * @note This function is only work for P1 element.
+   */
+  void
+  intersect_line_with_edge(std::vector<std::pair<double, Point<1>>> &t_vals,
+                           std::array<Point<2>, 2> element_vertices,
+                           dealii::Point<2>        r,
+                           dealii::Tensor<1, 2>    q,
+                           const double            tol);
+
+  /**
+   * This code is used to determine if the given line would intersect with the
+   * triangle Also perturb r if the FD stencil cutting boundary of the triangle.
+   *
+   * https://mathworld.wolfram.com/TriangleInterior.html
+   *
+   * @param[in] elements_vertices The array containing Eulerian location of
+   * vertices of the given element.
+   *
+   * @param[in] r The center location of the FD stencil`.
+   *
+   * @param[in] q a unit vector aligned with the coordinate axes.
+   *
+   * @param[in] tol a given tolerance for deciding if the intersection point is
+   * within the edge.
+   */
+  bool
+  if_line_intersect_with_triangle(
+    const std::array<Point<3>, 3> &element_vetices,
+    dealii::Point<3>              &r,
+    const dealii::Tensor<1, 3>    &q);
```

Comment:
I don't think we need this and the `intersect_line_with_` functions if we switch to using `std_cxx17::optional`. That will return a false value if there is no intersection and true otherwise.

Reply:
Thanks David for the comment. I agree.
One thing I am not sure about this change is In IBAMR we loop over elements and find corresponding stencil box to calculate the intersection point. But in some very small mfac case, we are trying to calculate intersection points for lots of triangles that within the same stencil box which are not cutted by stencil. I am not totally sure how we gonna search for intersections, but if we do it element by element do we need a function to judge if the stencil gonna cut the elements before directly calculating intersection? Because judging if an element would be cutted by stencil seems much cheaper than directly calculating elements. Please don't get me wrong I don't have problems with remove this. Let me know what you think.

Reply:
> One thing I am not sure about this change is In IBAMR we loop over elements and find corresponding stencil box to calculate the intersection point.

This still needs to be written, but the tentative plan for `PatchIntersectionMap` (I'll add this to the road map) is that we can 
1. get the physical-space bounding box for an element
2. convert it into an index-space bounding box
3. check just the stencils (defined as pairs of cell centers) which start or end in the index-space bounding box
which is pretty similar. 

> I am not totally sure how we gonna search for intersections in fiddle, but if we do it element by element do we need a function to judge if the stencil gonna cut the elements before directly calculating intersection?

Returning an `optional<double>` object lets the function which computes where the intersection should be do both - it either returns (as a convex combination coefficient) the location of the intersection or returns a logically false value.